ports/mimxrt10xx/common-hal/busio/UART.c

```diff
@@ -79,126 +81,178 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
-    const uint32_t tx_count = sizeof(mcu_uart_tx_list) / sizeof(mcu_periph_obj_t);
+    // We are transmitting one direction if one pin is NULL and the other isn't.
+    bool is_onedirection = (rx != NULL) != (tx != NULL);
+    bool uart_taken = false;
 
-    for (uint32_t i = 0; i < rx_count; ++i) {
-        if (mcu_uart_rx_list[i].pin != rx)
-            continue;
+    const uint32_t rx_count = MP_ARRAY_SIZE(mcu_uart_rx_list);
+    const uint32_t tx_count = MP_ARRAY_SIZE(mcu_uart_tx_list);
 
-        for (uint32_t j = 0; j < tx_count; ++j) {
-            if (mcu_uart_tx_list[j].pin != tx)
+    // RX loop handles rx only, or both rx and tx
+    if (rx != NULL) {
+        for (uint32_t i = 0; i < rx_count; ++i) {
+            if (mcu_uart_rx_list[i].pin != rx) {
                 continue;
-
-            if (mcu_uart_tx_list[j].bank_idx != mcu_uart_rx_list[i].bank_idx)
+            }
+            // If TX is on, keep looking, else stop
+            if (tx != NULL) {
+                for (uint32_t j = 0; j < tx_count; ++j) {
+                    if (mcu_uart_tx_list[j].pin != tx || 
+                        mcu_uart_tx_list[j].bank_idx != mcu_uart_rx_list[i].bank_idx) {
+                        continue;
+                    }
+                    // If UART is taken, break (pins never have >1 periph)
+                    if (reserved_uart[mcu_uart_rx_list[i].bank_idx - 1]) {
+                        uart_taken = true;
+                        break;
+                    }
+                    self->rx = &mcu_uart_rx_list[i];
+                    self->tx = &mcu_uart_tx_list[j];
+                    break;
+                }
+                if (self->tx != NULL || uart_taken) {
+                    break;
+                }
+            } else {
+                if (reserved_uart[mcu_uart_rx_list[i].bank_idx - 1]) {
+                    uart_taken = true;
+                    break;
+                }
+                self->rx = &mcu_uart_rx_list[i];
+            }
+        }
+    } else if (tx != NULL) {
+        // TX only case
+        for (uint32_t i = 0; i < tx_count; ++i) {
+            if (mcu_uart_tx_list[i].pin != tx) {
                 continue;
-
-            self->rx_pin = &mcu_uart_rx_list[i];
-            self->tx_pin = &mcu_uart_tx_list[j];
-
+            }
+            if (reserved_uart[mcu_uart_tx_list[i].bank_idx - 1]) {
+                uart_taken = true;
+                break;
+            }
+            self->tx = &mcu_uart_tx_list[i];
             break;
         }
+    } else {
+        mp_raise_ValueError(translate("Supply at least one UART pin"));
     }
 
-    if(self->rx_pin == NULL || self->tx_pin == NULL) {
+    if (uart_taken) {
+        mp_raise_RuntimeError(translate("Hardware in use, try alternative pins"));
+    }
+
+    if(self->rx == NULL && self->tx == NULL) {
         mp_raise_RuntimeError(translate("Invalid UART pin selection"));
```

Comment:
This should probably be a separate PR, but in all the other ports this is a `mp_raise_ValueError(translate("Invalid pins"));`, it matters because it breaks, for example, this code: https://learn.adafruit.com/circuitpython-essentials/circuitpython-uart-serial#wheres-my-uart-2985036-14